File: src/audioWfc/wfc/constraints/PerfectCadenceSoftConstraint.java
package audioWfc.wfc.constraints;

import audioWfc.wfc.constraints.concepts.GenericCadenceSoftConstraint;
import audioWfc.wfc.grabbers.ChordQualityConstantGrabber;
import audioWfc.wfc.grabbers.Grabber;
import audioWfc.wfc.grabbers.IntegerConstantGrabber;
import audioWfc.musicTheory.Key;
import audioWfc.musicTheory.chords.ChordQuality;

public class PerfectCadenceSoftConstraint extends GenericCadenceSoftConstraint {

    public PerfectCadenceSoftConstraint(double factor, Grabber<Key> keyGrabber){
        super(factor, keyGrabber, new IntegerConstantGrabber(7), new ChordQualityConstantGrabber(ChordQuality.MAJOR),
                new IntegerConstantGrabber(0), new ChordQualityConstantGrabber(ChordQuality.MAJOR));
    }

    @Override
    public String name() {
        return "Perfect cadences";
    }

    @Override
    public String configText() {
        return keyGrabber.configText() + ", Bonus: " + factor;
    }
}

File: vite-project/src/wfc/constraints/cadences/PerfectCadenceSoftConstraint.ts

import { ConcreteChordQuality } from "../../../music_theory/Chord"
import { GenericCadenceSoftConstraint } from "./GenericCadenceSoftConstraint"
import { Grabber } from "../../Grabber"
import { constantGrabber } from "../../grabbers/constantGrabbers"
import { MusicalKey } from "../../../music_theory/MusicalKey"
import { chordConstraintTypeToName } from "../constraintUtils"

export class PerfectCadenceSoftConstraint extends GenericCadenceSoftConstraint {
	name = chordConstraintTypeToName.get("PerfectCadenceSoftConstraint") as string
	configText = () => `Key: ${this.keyGrabber.configText()}, Bonus: ${this.bonus}`
	constructor(bonus: number, keyGrabber: Grabber<MusicalKey>) {
		super(bonus, constantGrabber(7), constantGrabber(0), constantGrabber("major" as ConcreteChordQuality), constantGrabber("major" as ConcreteChordQuality), keyGrabber)
	}
}